Use nested blocks for quotes

ellatrix opened the pull request:
## Description

Related: #5265.

This branch changes both quotes block so that they use nested blocks for the content. This enables the user to add content other than paragraphs and simplifies the block registration.

~~WIP, to do:~~

* ~~Preserve citation when unwrapping the blocks.~~
* ~~Make paste work.~~
* ~~Double enter should exit the quote.~~ Should this be a general nested block behaviour?
* ~~Update demo content.~~

## How Has This Been Tested?
* Ensure old quotes still work (try demo content).
* Ensure transformations work.
* Try the shortcut: `> + SPACE` both with and without existing content.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

iseulde commented on blocks/api/parser.js:
```diff
@@ -281,7 +281,8 @@ export function createBlockWithFallback( blockNode ) {
 			block.isValid = true;
-			block.attributes = attributesParsedWithDeprecatedVersion;
+			block.attributes = omit( attributesParsedWithDeprecatedVersion, '_innerBlocks' );
+			block.innerBlocks = attributesParsedWithDeprecatedVersion._innerBlocks || block.innerBlocks;
```
This is ugly, **but** I didn't want to change the deprecation API for this, and I suspect this will be a very rare thing to do. Might still want to change it though.

youknowriad replied:
A similar issue is being resolved in #5932 Let's try to find the best approach there.

iseulde replied:
Rebased.

youknowriad commented on blocks/rich-text/README.md:
```diff
@@ -31,9 +31,9 @@ a traditional `input` field, usually when the user exits the field.
 *Optional.* By default, a line break will be inserted on <kbd>Enter</kbd>. If the editable field can contain multiple paragraphs, this property can be set to `p` to create new paragraphs on <kbd>Enter</kbd>.
 
-### `onSplit( before: Array, after: Array, ...blocks: Object ): Function`
+### `onSplit( after: Array, ...blocks: Object ): Function`
```
Can you elaborate on how this works, it's a bit unclear?

Before, the HTML was splitted between two similar formats (after, and before) but right now it seems that the HTML before the split position is transformed to a block before calling the function (If I understand properly)? How do you decide which block to use since in the RichText component we don't really know the parent block? 

Alson If this is the case, why can't we just use `onReplace` and replace the `after` with a block as well.

youknowriad replied:
Also, it seems that we need to provide a deprecation message in any case.

iseulde replied:
> Before, the HTML was splitted between two similar formats (after, and before) but right now it seems that the HTML before the split position is transformed to a block before calling the function (If I understand properly)?

No, since the "before" content always lands in the some RichText area, we can do this as part of RichText. I don't thing there's a need to pass it to the block registration, which would only set attributes on the same block affecting the same RichText instance, which already has the correct content. So there's no need to pass as RichText is already setting it, currently the content is being set twice.

Good point about the deprecation message. Me might actually want to improve the shape of the arguments here as it's a bit of a mess. Maybe `{ blocks, afterValue }` or something similar.

iseulde replied:
> Alson If this is the case, why can't we just use onReplace and replace the after with a block as well.

Not sure I'm following here. The block needs to determine what to do with the split off value, which is arguably always a new block with the content set at the same attribute as the "before" value (the value that stays). The problem is that RichText should not be aware of the shape of the block's attributes...

youknowriad replied:
Don't worry about this comment ;) I thought initially you were doing some magic (transforming to blocks) but it's not the case. So my issue right now is more https://github.com/WordPress/gutenberg/pull/6054#discussion_r180326951

youknowriad commented on blocks/rich-text/index.js:
```diff
@@ -774,15 +778,15 @@ export class RichText extends Component {
 	restoreContentAndSplit( before, after, blocks = [] ) {
-		this.updateContent();
-		this.props.onSplit( before, after, ...blocks );
+		this.setContent( before );
```
Oh I see, so you apply the content directly beforing calling the `onSplit`. I believe we avoided this until now because the block can decide to not split and don't support the `Enter` behavior. If the block don't provide the `onSplit` it's better to keep content as is instead of removing the part that's after the split point. How do you handle this?

iseulde replied:
If there's no `onSplit` prop, `restoreContentAndSplit` will never be called, just like before? `updateContent` is being replaced by `setContent` here, because the only thing we want to do is set the content.

iseulde replied:
There was actually an inconsistency where some use called `setContent` before `restoreContentAndSplit`.

youknowriad replied:
👍 Sounds good to me. Let's try to build a good deprecation strategy though

iseulde replied:
Will do :) I'll ponder a bit over maybe using `onReplace`... would be much simpler, but right now `onSplit` is also used as an indicator whether splitting is possible as you mentioned, whereas onReplace is used for other things to. One solution to both problems might be to have something like this:

```
onSplit: ( value ) => createBlock ( name, {
  content: value,
}
```

So here the block provided RichText with the info it needs to split the block, and then we can use `onReplace` with `[ ...blocks, onSplit( value ) ]`.

But one problem with this is that we currently insert an empty paragraph for the list block if `blocks` is empty...

youknowriad replied:
Also maybe `onSplit` always create a `defaultBlock` for the "after" content in which case, we can replace it entirely?

iseulde replied:
This need will go away though once we use `innerBlocks` for the list block too...

iseulde replied:
Atm I'm tempted to continue passing the `before` value, but remove the `setAttributes`, then revise the whole function once we work on lists.

youknowriad replied:
Fine to leave for another refactor though 👍

iseulde replied:
Cool, let's take care of this when we handle lists.

youknowriad commented on editor/components/block-list/block.js:
```diff
@@ -650,10 +649,21 @@ const applyWithDispatch = withDispatch( ( dispatch, ownProps ) => {
+				insertBlocks( blocks, orderOfRoot + 1, rootUIDOfRoot );
+			} else {
+				insertBlocks( blocks, order + 1, rootUID );
+			}
```
Can you explain the need of this separate handler and its logic, not obvious at first sight :)

iseulde replied:
Oh sure! Forgot to comment. It's so escape the nested block list on double enter.

youknowriad replied:
A small e2e test for the double enter behavior might be good. It's something that can regress easily without noticing.

youknowriad commented on editor/components/block-settings-menu/block-unwrap.js:
```diff
@@ -0,0 +1,69 @@
+/**
```
Can we leave unwrapping to a separate PR. Not saying we shouldn't do it, but it seems better (at least for me) to discuss separately?

iseulde replied:
Sure, but how do I take care of the block transforms in this PR? Or should I only add the unwrapping if the block has specified a transform (vs always offering unwrap if the block has innerBlocks)?

youknowriad replied:
This is a new feature in this PR right? Is it something needed to replace an existing feature? I suspect it's here to replace some transforms that are not possible in the nested approach? Maybe we don't care that much about these transforms at the moment?

iseulde replied:
Yes, it's to take care of quote => paragraph/heading. See https://github.com/WordPress/gutenberg/pull/6054/files#diff-cda27cec5557c5b032f2b19c44309a1dL106. I added the unwrapping to show it's easier to take care of these kind of transformations. But I agree this should maybe be discussed more. Are you fine with dropping these transformations then at the moment?

iseulde replied:
Okay I removed this for now.

youknowriad reviewed (commented):
Nice work on this PR. I left some comments but it's great to see how these blocks become way simpler now.

afercia commented:
Sorry I haven't followed so much so maybe I'm missing something, in that case please ignore my question. Could you please specify what you mean by:
> This enables the user to add content other than paragraphs

Worth noting the `blockquote` element is meant to be used for... quotes, that is textual content which typically needs just paragraphs and, optionally, a footer or cite element for the attribution.

youknowriad commented:
@afercia With #5452 we'll be able to limit the allowed block types for nested blocks per parent block. Also, people wanted to be able to add headings and lists in quotes.

afercia commented:
@youknowriad that was my concern...

> people wanted to be able to add headings and lists in quotes

That sounds totally weird to me. Quotes are not meant to be used for visual purposes. A `blockquote` element is the representation of a quote, nothing more, nothing less. Though headings and lists in a `blockquote` seems to be valid HTML, seems to me they're not semantic at all. WordPress aims to be a `semantic publishing platform` and one of its first goals is to produce rich, semantic, HTML. I'd suggest to wait a bit with this implementation and consider what we're really doing here from a semantics perspective.

In the meantime, I'd suggest to have a look at the specification and all the HTML examples there: https://www.w3.org/TR/html52/grouping-content.html#the-blockquote-element

iseulde commented:
@afercia It should be possible to use lists and other blocks in a quote block. I should be able to quote this issue:

> WIP, to do:
> 
> * ~~Preserve citation when unwrapping the blocks.~~
> * ~~Make paste work.~~
> * ~~Double enter should exit the quote.~~ Should this be a general nested block behaviour?
> * ~~Update demo content.~~

This is also possible currently in the classic editor and elsewhere (e.g. Markdown).

This PR is also about more than just allowing other blocks in the quote. It's also about simplifying the way the nested paragraphs are implemented. If we end up deciding only paragraphs should be allowed in quotes, we can restrict the blocks that are allowed in a quote.

afercia commented:
@iseulde sorry that's not a quote in a semantic sense as described in the HTML specification :) That's something more similar to a reply with the original message attached and indented as email clients do with `>`.

> Also, people wanted to be able to add headings and lists in quotes.

It would be great to have access to this feedback and see what are the real use cases users wish to have.

iseulde commented:
> sorry that's not a quote in a semantic sense as described in the HTML specification :)

Why not? Could you elaborate?

> The blockquote element represents content that is quoted from another source, optionally with a citation which must be within a footer or cite element, and optionally with in-line changes such as annotations and abbreviations.

I should be able to quote content form another source, including its semantic structure.

afercia commented:
@iseulde That's not to say a block for your use case can't be done but then it shouldn't use a `blockquote` element :) that would be just a representation of a "conversation". From the specification:

> The blockquote element represents content that is quoted from another source, optionally with a citation which must be within a footer or cite element, and optionally with in-line changes such as annotations and abbreviations.

I'd point out the part **"quoted from another source"**.

See also the NOTE about how to represent a conversation:
> Examples of how to represent a conversation are shown in a later section; it is not appropriate to use the cite and blockquote elements for this purpose.

Examples here: https://www.w3.org/TR/html52/common-idioms-without-dedicated-elements.html#conversations

youknowriad commented:
@afercia According to the link you shared. The content of the blockquote allows any "Flow Content" element which means any of these elements: https://www.w3.org/TR/html52/dom.html#flow-content-2

<img width="805" alt="screen shot 2018-04-10 at 10 39 47" src="https://user-images.githubusercontent.com/272444/38549247-875d9b1c-3cab-11e8-878a-edf7726c12e5.png">

This is also confirmed by other sources like https://developer.mozilla.org/en-US/docs/Web/HTML/Element/blockquote

rianrietveld commented:
Commenting at the request of @afercia:
According to the [specs at MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/blockquote), a quote can contain flow content, so headings and list and such are allowed.

Looking at the actual meaning and good use of a blockquote, resticting this to only a `p`, `footer`, `cite` would be more relevant, but also the [W3C specs](https://w3c.github.io/html/grouping-content.html#the-blockquote-element) allow flow content.

But for a CMS like WordPress where a blockquote is used already in so many ways, forcing this on content managers, out of pure semantics, seem a bridge too far to me, also for backwards compatibility of text in older posts.

aristath commented:
While the purist in me agrees 100% that quotes are - and should be - quotes and nothing but quotes, in real-life I've seen lots of sites using quotes to make something stand out. I assume the reason for this is because lots of themes just make quotes pretty and visually appealing. Is it wrong? definitely. Do people still use it just to make a visual impression? Unfortunately yes.

afercia commented:
@youknowriad as I've previously commented, yeah it seems to be valid HTML. However, my point is about semantics. See the above comments and the NOTE on the specification.

afercia commented:
> I've seen lots of sites using quotes to make something stand out

That's a perfect use case for another block type that doesn't use the `blockquote` element 🙂

iseulde commented:
I'm not representing conversion. I want to quote a source.

E.g.

> Primary aromatic amines are used as a starting material for the manufacture of azo dyes. It reacts with nitrous acid to form diazonium salt, which can undergo coupling reaction to form an azo compound. As azo-compounds are highly coloured, they are widely used in dyeing industries, such as:
> 
> * Methyl orange
> * Direct brown 138
> * Sunset yellow FCF
> * Ponceau
>
> *— Some article*

afercia commented:
@iseulde yeah, seems a valid edge case. Question: how many users will abuse this block and use it just for visual purposes?

pento commented:
So help me, I will turn this car around if you kids don't stop arguing. 😉

iseulde commented:
The block clearly says it's a quote, we can even elaborate in the description. You can misuse any block for that. We have to trust the user somewhere. It's more likely they will misuse headings for big text...

Not that it is a valid argument in every situation, but we also allowed this content in the classic editor, it's semantic and valid HTML.

aristath commented:
> Question: how many users will abuse this block and use it just for visual purposes?

I'm guessing that depends on the theme they use, but I don't think it will be any different than what we have now... Those that abuse it will continue to abuse it, and if we add a description that elaborates a bit more like @iseulde suggested maybe some of them will understand and stop. I don't think there's anything more that can - or should - be done.

afercia commented:
> So help me, I will turn this car around if you kids don't stop arguing.

Sorry I'm not a native English speaker, so I don't get the irony, which I found inappropriate in this case anyway, since we're a global community and everyone should strive to make his communication as effective and inclusive as possible.

> it's semantic and valid HTML.

Valid: yes. Semantic: not so sure.

> Those that abuse it will continue to abuse it

Yeah, but on the other side nothing stops us from trying to improve things. I'd rather stick to the specification and better semantics and use a `blockquote` element for real quotes. If users want a block to be used for content that "stands out", then the purpose is completely different and shouldn't be rendered with a `blockquote` element.

iseulde commented:
> Yeah, but on the other side nothing stops us from trying to improve things. I'd rather stick to the specification and better semantics and use a blockquote element for real quotes. If users want a block to be used for content that "stands out", then the purpose is completely different and shouldn't be rendered with a blockquote element.

We are sticking to the spec, and it's up to the user to use them for real quotes. There's nothing that we can do about that, not in current master, not in this branch. It's so easy right now to abuse the quote block as well even if it's limited to an infinite amount of paragraphs:

> This is NOT a quote. I just want to make it big.

I didn't even need lists for that.

So only allowing some more blocks inside a quote will not make this abuse much worse, but it *will* allow you to create perfectly valid and semantic quotes for a bigger variety of sources.

iseulde commented:
Note: I think this is a fine discussion to have and I'm more than happy to explain the motivations behind this PR. As I said, it's about more than just allowing more block types in quotes. If we really want to, we can always restrict it later, though I don't share this opinion.

afercia commented:
> So only allowing some more blocks inside a quote will not make this abuse much worse, but it will allow you to create perfectly valid and semantic quotes for a bigger variety of sources.

I kindly disagree 🙂 What this PR is actually doing, is allowing users to greatly abuse of a `blockquote` element, and users will do, as they always do. The fact the classic editor is already allowing it, is not the best argumentation 🙂 I'd rather consider a "stand out" block, where users can put any HTML they want.

Also, my previous point still stands:

>> people wanted to be able to add headings and lists in quotes.

> It would be great to have access to this feedback and see what are the real use cases users wish to have.

Can we have access to the mentioned feedback and see what are the real use cases people mentioned?

iseulde commented:
> The fact the classic editor is already allowing it, is not the best argumentation

No, that mere fact alone is not, but we're also enabling near seamless conversion from old post content, and the fact the HTML is valid, can be semantic, and has really good use cases (gave an example) still remains. We also know that there will be "abuse" before and after, I don't think this is a huge game changer for people wanting to abuse semantic HTML tags. You can abuse current quotes, headings, inline tags, code, preformatted etc. with the same reasons.

I do get it though... the line is hard to draw and e.g. for the table block I think we will only ever support text and inline tags because adding virtually anything else would be better done in dedicated layout blocks (no good use cases). This is a whole other level though, IMHO.

youknowriad commented on blocks/inner-blocks/index.js:
```diff
@@ -4,13 +4,13 @@
 
 function InnerBlocks( { BlockList, layouts } ) {
-	return <BlockList layouts={ layouts } />;
+	return BlockList ? <BlockList layouts={ layouts } /> : null;
```
In which case we need these checks?

iseulde replied:
For tests, they will error.

afercia commented:
Yeah I agree the line is hard to draw. Fair enough, I've expressed my concerns 🙂 I'd recommend to restrict the allowed block types as much as possible. Worth noting WordPress already does this in some cases, not only for HTML elements but also for some attributes, for better semantics and other reasons too.

SuperGeniusZeb commented:
Just thought I would drop my opinion here:

I think it makes sense to allow lists and headings in blockquotes because it is fully valid (and semantic) HTML to have a blockquote that includes those things. Quotes are not limited to paragraphs because a source does not always contain just paragraphs. I think technically, any semantic blocks should be allowed to be nested in the Quote block. So the following core blocks should be allowed:
- Code
- Heading
- List
- Paragraph
- Preformatted
- Subhead
- Verse

The following could be allowed, but I do not think they are really necessary, and it may be a bad idea to allow them:
- Audio (Used in posts far more rarely than images, and unlikely to be needed in a quote, though I could see a use case if you wanted to semantically quote someone using a recording of their voice, rather than text. Most people would probably just use an Audio block by itself, but nesting it in a Quote block would technically be more semantically correct in some cases.)
- Custom HTML (If you are using some script or shortcode or something to pull in a quote, or are in some other edge-case situation, this block would be useful, but I do not think it is really necessary, and it feels like it could open a can of worms.)
- Image (Almost never needed in a quote unless the image is a vital part of the content, which is really rare, though it may happen in instances where you are trying to exactly quote an entire forum post or social media post or something like that.)
- Quote (Technically, you can quote someone quoting someone else. This reminds me of forum posts, which is probably the most obvious use for this. I can not see this being used much, but I also can not see this being abused much either, because why would you ever nest a blockquote for just styling reasons?)
- Separator (Technically has some semantic meaning, but unlikely to ever be needed or wanted.)
- Table (I do not think most people like using tables in the first place, and I doubt quoting them is going to happen very often, though it is probably one of the more likely to be allowed in this group.)
- Video (Like the Audio block use case, this is technically possible, but very unlikely to be used. Most of the time, you would probably just use a Video block with no nesting.)
- Embed blocks (See reasoning for Audio and Video blocks.)

The following would not be allowed:
- Button (I can not see any use case where you would actually need to quote a button.)
- Classic (This block is intended for legacy transitional use only. Its content can always just be represented using standard blocks.)
- Columns (This is used only for layout and styling. It has no semantic meaning as far as I am aware.)
- Cover Image (Especially if the block becomes one that holds nested blocks, at which point it would become a style-only block that does not add any semantic meaning and therefore definitely does not belong in a blockquote.)
- Gallery (No one is going to quote a gallery because a gallery adds nothing to the semantic content of a post in almost every situation.)
- More (It is not even technically part of the content of an article... it is just a marker to tell where to cut it off when displaying it in archives and other post lists.)
- Pullquote (This is basically just a highlight of content that exists in another place in an article, so I would not really consider it as an actual part of an article. It is more of a stylistic thing in my opinion, so it does not belong in a Quote block.)
- Shortcode (There may be some odd edge case that I do not know about, but I doubt this would ever be needed.)
- Text Columns (Same reasons as the Columns block.)
- Categories, Latest Posts, and any other "widget" blocks (They are almost never actually part of the content of an article and would not even work in the context of another website quoting the content.)

Additionally, I think that headings in blockquotes should not be shown in the Table of Contents, as they are not the headings of the post itself, but of the sources that are being quoted.

iseulde commented:
Investigating error on enter.

iseulde commented:
Added an e2e test for the double enter behaviour, which works locally but fail on Travis. Unsure what is going on.

iseulde commented:
Rebased (and squashed).

youknowriad commented on test/e2e/support/utils.js:
```diff
@@ -58,3 +58,16 @@ export async function newDesktopBrowserPage() {
+	await button.click( 'button' );
+}
+
+export async function getHTMLFromCodeEditor() {
```
Nice refactoring, we talked about having a special jest asserter, but this is nice as well.

youknowriad reviewed (commented):
LGTM 👍 Nice job

ellatrix committed:
Use nested blocks for quotes
81aec55
Don't use rawHandler for migration
188e4a8
Make paste work
86770b7
Adjust e2e, multi selection does not work across nested levels
23027a7
Preserve citation when unwrapping the blocks
188150c
Update demo content
2900215
ENTER twice should remove nesting
f69f171
Avoid updateContent call on split
a54b12d
Restore before arg in onSplit, refactor later
bf8c171
Handle block (un)wrapping in a separate PR
8929b5f
Add e2e test for nested block splitting
ef09336
Commits on Apr 12, 2018
Use moveBlockToPosition
6662c36
Fix TinyMCE error on double enter split
6ab8eea
Fix split and merge
ec7f609

iseulde commented:
Rebased. I'm going to merge, we can limit the blocks allowed once the other PR is merged.

SuperGeniusZeb commented:
Should this be removed from the 2.8 milestone since it got reverted?

mcsf commented:
I've instead added _both_ this and its revert to the milestone, so we have history of this work.

iseulde commented:
Thanks! :)
